File: docs/3-guides/developer/api-examples/python/query_findings.py
"""
Example: Query audit findings using Python requests library.

This script demonstrates how to query the SQL Audit Service API
to retrieve critical findings using the Python requests library.

Prerequisites:
  - Audit service running at localhost:8090
  - requests library: pip install requests

Usage:
  python query_findings.py
"""

import requests
from datetime import datetime, timedelta
from typing import List, Dict, Optional


BASE_URL = "http://localhost:8090/api/v1"


def query_critical_findings(limit: int = 10) -> Dict:
    """
    Query recent CRITICAL audit findings.

    Args:
        limit: Maximum number of findings to retrieve

    Returns:
        Dictionary containing findings and pagination info
    """
    url = f"{BASE_URL}/audits"
    params = {
        "riskLevel": "CRITICAL",
        "size": limit,
        "sort": "createdAt,desc"
    }

    try:
        response = requests.get(url, params=params)
        response.raise_for_status()

        data = response.json()

        print(f"Total CRITICAL findings: {data['totalElements']}")
        print(f"Page {data['number'] + 1} of {data['totalPages']}")
        print("\nFindings:")

        for finding in data['content']:
            print("-" * 60)
            print(f"ID: {finding['id']}")
            print(f"SQL: {finding['sql']}")
            print(f"Risk Score: {finding['riskScore']}")
            print(f"Checker: {finding['checkerId']}")
            print(f"Message: {finding['message']}")
            print(f"Recommendation: {finding['recommendation']}")

        return data

    except requests.exceptions.RequestException as e:
        print(f"Error querying audit findings: {e}")
        return {}


def query_with_filters(
    sql_id: Optional[str] = None,
    risk_level: Optional[str] = None,
    start_time: Optional[datetime] = None,
    end_time: Optional[datetime] = None,
    page: int = 0,
    size: int = 20
) -> Dict:
    """
    Query audit findings with custom filters.

    Args:
        sql_id: Filter by SQL identifier
        risk_level: Filter by risk level (CRITICAL, HIGH, MEDIUM, LOW)
        start_time: Start time for time range filter
        end_time: End time for time range filter
        page: Page number (zero-based)
        size: Page size

    Returns:
        Dictionary containing filtered findings
    """
    url = f"{BASE_URL}/audits"
    params = {
        "page": page,
        "size": size
    }

    if sql_id:
        params["sqlId"] = sql_id
    if risk_level:
        params["riskLevel"] = risk_level
    if start_time:
        params["startTime"] = start_time.isoformat()
    if end_time:
        params["endTime"] = end_time.isoformat()

    response = requests.get(url, params=params)
    response.raise_for_status()

    data = response.json()
    print(f"Filtered results: {data['totalElements']} findings")

    return data


def get_audit_by_id(report_id: str) -> Optional[Dict]:
    """
    Get a specific audit report by ID.

    Args:
        report_id: Audit report identifier

    Returns:
        Audit report dictionary or None if not found
    """
    url = f"{BASE_URL}/audits/{report_id}"

    try:
        response = requests.get(url)
        response.raise_for_status()

        report = response.json()

        print(f"Audit Report: {report['id']}")
        print(f"SQL: {report['sql']}")
        print(f"Risk Score: {report['riskScore']}")
        print(f"Metadata: {report.get('metadata', {})}")

        return report

    except requests.exceptions.HTTPError as e:
        if e.response.status_code == 404:
            print(f"Audit report not found: {report_id}")
        else:
            print(f"Error retrieving audit report: {e}")
        return None


def get_dashboard_stats() -> Dict:
    """
    Get dashboard statistics.

    Returns:
        Dictionary containing dashboard statistics
    """
    url = f"{BASE_URL}/statistics/dashboard"

    response = requests.get(url)
    response.raise_for_status()

    stats = response.json()

    print("Dashboard Statistics:")
    print(f"Total Findings: {stats['totalFindings']}")
    print(f"Critical: {stats['criticalCount']}")
    print(f"High: {stats['highCount']}")
    print(f"Medium: {stats['mediumCount']}")
    print(f"Low: {stats['lowCount']}")

    if 'topRiskySql' in stats:
        print("\nTop Risky SQL Patterns:")
        for sql in stats['topRiskySql'][:5]:
            print(f"  - Score {sql['riskScore']}: {sql['sql'][:50]}...")

    return stats


def get_trend_stats(
    start_time: datetime,
    end_time: datetime,
    granularity: str = "DAY"
) -> List[Dict]:
    """
    Get trend statistics over time.

    Args:
        start_time: Start time for trend data
        end_time: End time for trend data
        granularity: Time granularity (HOUR, DAY, WEEK, MONTH)

    Returns:
        List of trend data points
    """
    url = f"{BASE_URL}/statistics/trends"
    params = {
        "startTime": start_time.isoformat(),
        "endTime": end_time.isoformat(),
        "granularity": granularity
    }

    response = requests.get(url, params=params)
    response.raise_for_status()

    trends = response.json()

    print(f"Trend data ({granularity}):")
    for point in trends:
        print(f"  {point['date']}: {point['count']} findings "
              f"({point.get('criticalCount', 0)} critical)")

    return trends


def main():
    """Main function demonstrating various API calls."""
    print("=== Querying Critical Findings ===")
    query_critical_findings(limit=5)

    print("\n=== Dashboard Statistics ===")
    get_dashboard_stats()

    print("\n=== Trend Statistics (Last 7 Days) ===")
    end_time = datetime.now()
    start_time = end_time - timedelta(days=7)
    get_trend_stats(start_time, end_time, granularity="DAY")

    print("\n=== Filtered Query ===")
    query_with_filters(
        risk_level="HIGH",
        start_time=start_time,
        size=10
    )


if __name__ == "__main__":
    main()

File: docs/3-guides/developer/api-examples/javascript/queryFindings.js
/**
 * Example: Query audit findings using fetch API (Node.js or Browser).
 *
 * This script demonstrates how to query the SQL Audit Service API
 * using the native fetch API.
 *
 * Prerequisites:
 *   - Audit service running at localhost:8090
 *   - Node.js 18+ (for native fetch support)
 *
 * Usage:
 *   node queryFindings.js
 */

const BASE_URL = 'http://localhost:8090/api/v1';

/**
 * Query recent CRITICAL audit findings.
 */
async function queryCriticalFindings(limit = 10) {
    const url = new URL(`${BASE_URL}/audits`);
    url.searchParams.append('riskLevel', 'CRITICAL');
    url.searchParams.append('size', limit.toString());
    url.searchParams.append('sort', 'createdAt,desc');

    try {
        const response = await fetch(url);

        if (!response.ok) {
            throw new Error(`HTTP error! status: ${response.status}`);
        }

        const data = await response.json();

        console.log(`Total CRITICAL findings: ${data.totalElements}`);
        console.log(`Page ${data.number + 1} of ${data.totalPages}`);
        console.log('\nFindings:');

        data.content.forEach(finding => {
            console.log('----------------------------------------');
            console.log(`ID: ${finding.id}`);
            console.log(`SQL: ${finding.sql}`);
            console.log(`Risk Score: ${finding.riskScore}`);
            console.log(`Checker: ${finding.checkerId}`);
            console.log(`Message: ${finding.message}`);
            console.log(`Recommendation: ${finding.recommendation}`);
        });

        return data;
    } catch (error) {
        console.error('Error querying audit findings:', error.message);
        throw error;
    }
}

/**
 * Query audit findings with custom filters.
 */
async function queryWithFilters(options = {}) {
    const {
        sqlId,
        riskLevel,
        startTime,
        endTime,
        page = 0,
        size = 20
    } = options;

    const url = new URL(`${BASE_URL}/audits`);
    url.searchParams.append('page', page.toString());
    url.searchParams.append('size', size.toString());

    if (sqlId) url.searchParams.append('sqlId', sqlId);
    if (riskLevel) url.searchParams.append('riskLevel', riskLevel);
    if (startTime) url.searchParams.append('startTime', startTime);
    if (endTime) url.searchParams.append('endTime', endTime);

    const response = await fetch(url);

    if (!response.ok) {
        throw new Error(`HTTP error! status: ${response.status}`);
    }

    const data = await response.json();
    console.log(`Filtered results: ${data.totalElements} findings`);

    return data;
}

/**
 * Get a specific audit report by ID.
 */
async function getAuditById(reportId) {
    const url = `${BASE_URL}/audits/${reportId}`;

    try {
        const response = await fetch(url);

        if (!response.ok) {
            if (response.status === 404) {
                console.log(`Audit report not found: ${reportId}`);
                return null;
            }
            throw new Error(`HTTP error! status: ${response.status}`);
        }

        const report = await response.json();

        console.log(`Audit Report: ${report.id}`);
        console.log(`SQL: ${report.sql}`);
        console.log(`Risk Score: ${report.riskScore}`);
        console.log(`Metadata:`, report.metadata);

        return report;
    } catch (error) {
        console.error('Error retrieving audit report:', error.message);
        return null;
    }
}

/**
 * Get dashboard statistics.
 */
async function getDashboardStats() {
    const url = `${BASE_URL}/statistics/dashboard`;

    const response = await fetch(url);

    if (!response.ok) {
        throw new Error(`HTTP error! status: ${response.status}`);
    }

    const stats = await response.json();

    console.log('Dashboard Statistics:');
    console.log(`Total Findings: ${stats.totalFindings}`);
    console.log(`Critical: ${stats.criticalCount}`);
    console.log(`High: ${stats.highCount}`);
    console.log(`Medium: ${stats.mediumCount}`);
    console.log(`Low: ${stats.lowCount}`);

    if (stats.topRiskySql) {
        console.log('\nTop Risky SQL Patterns:');
        stats.topRiskySql.slice(0, 5).forEach(sql => {
            console.log(`  - Score ${sql.riskScore}: ${sql.sql.substring(0, 50)}...`);
        });
    }

    return stats;
}

/**
 * Get trend statistics over time.
 */
async function getTrendStats(startTime, endTime, granularity = 'DAY') {
    const url = new URL(`${BASE_URL}/statistics/trends`);
    url.searchParams.append('startTime', startTime);
    url.searchParams.append('endTime', endTime);
    url.searchParams.append('granularity', granularity);

    const response = await fetch(url);

    if (!response.ok) {
        throw new Error(`HTTP error! status: ${response.status}`);
    }

    const trends = await response.json();

    console.log(`Trend data (${granularity}):`);
    trends.forEach(point => {
        console.log(`  ${point.date}: ${point.count} findings ` +
                    `(${point.criticalCount || 0} critical)`);
    });

    return trends;
}

/**
 * Main function demonstrating various API calls.
 */
async function main() {
    try {
        console.log('=== Querying Critical Findings ===');
        await queryCriticalFindings(5);

        console.log('\n=== Dashboard Statistics ===');
        await getDashboardStats();

        console.log('\n=== Trend Statistics (Last 7 Days) ===');
        const endTime = new Date().toISOString();
        const startTime = new Date(Date.now() - 7 * 24 * 60 * 60 * 1000).toISOString();
        await getTrendStats(startTime, endTime, 'DAY');

        console.log('\n=== Filtered Query ===');
        await queryWithFilters({
            riskLevel: 'HIGH',
            startTime: startTime,
            size: 10
        });
    } catch (error) {
        console.error('Error in main:', error.message);
        process.exit(1);
    }
}

// Run main function if executed directly
if (require.main === module) {
    main();
}

// Export functions for use as a module
module.exports = {
    queryCriticalFindings,
    queryWithFilters,
    getAuditById,
    getDashboardStats,
    getTrendStats
};
